### ISSUE TITLE
Performance as is

### ISSUE THREAD

**Issue by kanonka** [BobMcDear/NN_CUDA#4]
Hi, in your article you mentioned performance as is to be 6.3 second the GPU.

I took your github code as is, compiled for the sm_86 target (Nvidia 3090), latest CUDA and drivers, and time it is taking is in minutes (like ~30 minutes). Compiled in release mode, Windows 10, MSVC++ 2022.

What am I doing wrong?

I also tried to play with block_size. When increased to 64 from 32, it's much faster (matter of seconds now), but calculated loss doesn't depend on number of epochs - it's always 0.02 no matter what. 

Something is wrong.

**Comment by BobMcDear** [comment]
Hello,

Yes, the code takes ~6.3 seconds on a RTX 2070. I have tested it on other GPUs, including the RTX 3090, and the training time is more or less constant. However, all experiments were conducted on Linux, and therefore Windows is most likely the source of the problem. Another user encountered [a similar phenomenon](https://github.com/BobMcDear/Neural-Network-CUDA/issues/2#issuecomment-1114318960) with WSL, and it might explain your issue. Have you run other CUDA/C++ programs on your system before, specifically [these official samples](https://github.com/nvidia/cuda-samples)? Also, is dual-booting with Linux an option? 

One should be careful when modifying the block size. Specifically, some values would yield erroneous results for certain operations, data shapes, etc. In this case, any block size greater than 32 would execute faster with no explicit errors, but the output would be incorrect, and the network would not train. That is why the number of epochs does not affect the loss, which should be 0.003 (after 100 epochs) and not 0.02. 

Thank you.

**Comment by kanonka** [comment]
Ok, found the reason:
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html

Copy from there for the function cudaMallocManaged:
* On Linux, the location chosen will be **device memory** [...]
* On Windows, the physical storage is always created in 'zero-copy' or **host memory**. [...]

I tried the remedy they recommended (setting CUDA_MANAGED_FORCE_DEVICE_ALLOC=1), but that immediately fails when trying to call any host function over device memory (as it should).

Basically, cudaMallocManaged is bad idea on Windows.

Oh, well, I was going to use Thrust and CuBLAS anyway to adapt your code...

**Comment by BobMcDear** [comment]
Thanks for the information. I have virtually no development experience in Windows, so I am out of my element, but it is strange that ```cudaMallocManaged``` is not compatible with Windows. An [official guide](https://developer.nvidia.com/blog/even-easier-introduction-cuda/) by Nvidia on ```cudaMallocManaged``` indicates it can be run on Windows, and after skimming through the documentation, it appears to me that, as long one does not omit ```cudaDeviceSynchronize```, communication between the device and host should be seamless. Again though, I have no knowledge of development on Windows and might be mistaken. 

Please let me know if this issue can be closed if you have no further concerns.

**Comment by kanonka** [comment]
Yes, issue can be closed.

**Comment by kanonka** [comment]
Just FYI (feel free to close the issue again after you read this):

I just went and did a dumb change: all your float arrays are replaced with thrust::device_vector<float>. No other changes.
This made it run for 1.7 sec (release mode), final loss = 0.003. Looks good to me, but is there any better way to ensure that training went ok and achieved correct state (instead of skipping everything)? My concern is timing - you mentioned that it's roughly ~6 sec on all cards, and 1.7 sec seems too fast (although RTX 3090 is much faster than 2070, but you saying that training time for this test case is more or less constant on various cards).

Another question is this: in release mode, if I enable fast math on device (i.e. switch -use-fast-math), no matter how many epochs (even 1), I'm getting NaN as final loss. Any idea why this could be? I didn't dive deep into the code yet, just trying to find if I can use your code as a starting point.

**Comment by BobMcDear** [comment]
Thrust should not boost speed, however, it vastly simplifies memory management, which appears to be knotty on Windows and is the cause of your problem.



> is there any better way to ensure that training went ok and achieved correct state

Under the ```test/``` folder are tests that compare the CPU and GPU implementations. If they pass and the loss is roughly 0.003, it can be established that the code is fine. 

> you mentioned that it's roughly ~6 sec on all cards, and 1.7 sec seems too fast

By "more or less constant," I meant that the speed disparity does not reach an order of magnitude, but ~6 seconds on the RTX 2070 and ~2 on the RTX 3090 is not surprising. I do not recall the precise execution time on the RTX 3090, but I do remember that it was about twice as fast as the RTX 2070.



> if I enable fast math on device (i.e. switch -use-fast-math), no matter how many epochs (even 1), I'm getting NaN as final loss

When the first argument of the ```powf``` function is negative and the fast version is used, it leads to a NaN. Hence, ```mse_forward_gpu``` engenders NaNs assuming ```inp[ind]-out[ind]``` is smaller than 0. To fix it, the absolute value of ```inp[ind]-out[ind]``` could be passed to ```powf```, or ```powf(inp[ind]-out[ind], 2)``` can be replaced with ```(inp[ind]-out[ind])*(inp[ind]-out[ind])```. Either way, bear in mind that ```mse_forward_gpu``` is utilized only for printing the loss and does not influence training, so the true loss is 0.003 regardless of whether the displayed loss is NaN or not.

Is this helpful?

**Comment by kanonka** [comment]
> Is this helpful?

Absolutely!
Thank you again!
